pkgpanda/util.py

```diff
@@ -95,11 +126,82 @@ def copy_directory(src_path, dst_path):
         # thrown at it.
         src = src_path.replace('/', '\\')
         dst = dst_path.replace('/', '\\')
-        subprocess.check_call(['cmd.exe', '/c', 'xcopy', src, dst, '/E', '/B', '/I'])
+        subprocess.check_call(['cmd.exe', '/c', 'xcopy', src, dst, '/E', '/B', '/I'], stdout=subprocess.DEVNULL)
     else:
         subprocess.check_call(['cp', '-r', src_path, dst_path])
 
 
+def link_file(src, dst):
+    """Create file links in an OS independent manner.
+
+    On Linux, this will create symbolic links for both files and directories.
+    On Windows, this will create hard links for files, and junctions (a.k.a symbolic links) for directories.
+
+    """
+    if is_windows:
+        if os.path.isdir(src):
+            # create a junction for directories
+            src = src.replace('/', '\\')
+            dst = dst.replace('/', '\\')
+            subprocess.check_call(['cmd.exe', '/c', 'mklink', '/J', dst, src], stdout=subprocess.DEVNULL)
+        else:
+            # create hard link for files
+            os.link(src, dst)
+
+    else:
+        os.symlink(src, dst)
+
+
+def _is_junction(path):
+    if not is_windows:
+        raise Exception("Calls to _is_junction() only supported on Windows")
+
+    # TODO(klueska): Investigate why we need this 'isdir()' check. As far as I
+    # can tell, the check below that relies on 'GetFileAttributesW()' will
+    # return 'True' in cases where the 'path' passed to it is a symbolic link for
+    # either a _file_ or a _directory_. Since a 'junction' is only a directory
+    # level concept, its possible that the 'isdir()' check is required to limit
+    # us to directories only, before we ever make the call to
+    # 'GetFileAttributesW()'. This should be confirmed though.
+    if not os.path.isdir(path):
+        return False
+
+    file_attribute_reparse_point = 0x0400
+    attributes = ctypes.windll.kernel32.GetFileAttributesW(path)
+    return (attributes & file_attribute_reparse_point) > 0
+
+
+def islink(path):
+    """Check if a file is a link to another file in an OS independent manner.
+
+    On Linux, this will check for symbolic links for both files and directories.
+    On Windows, this will check for hard links for files, and junctions (a.k.a symbolic links) for directories.
+
+    """
+    if is_windows:
+        if os.path.isdir(path):
+            return _is_junction(path)
+        elif os.stat(path).st_nlink > 1:
```

Comment:
There are only a few places where `islink` is used. In one, it is only used for directories. In the other, Windows is skipped because this code identifies a real file as a link. Given that we are not using Windows symbolic links (BTW any reason for that?), it seems that it would be OK to return False for all Windows files and simplify the places where `islink` is used.

Reply:
I think there's a case in `pkgpanda/__init__.py::symlink_tree()` where we use `islink()` on files:

```python
                if not os.path.isdir(dest_path) and not islink(dest_path):
```